File: PO/Project/JavaProject/src/Animals/Turtle.java
package Animals;

import Abstractions.*;
import com.google.gson.JsonObject;

import java.util.Random;
import java.awt.*;
import java.util.concurrent.ThreadLocalRandom;

public class Turtle extends Animal {

    public Turtle(World world, Point position) {
        super(2, 1, position, world);
        this.avatar = 'T';
        this.image_path = "Turtle.png";
        loadImage();
    }

    @Override
    public int action() {
        Random rand = new Random();
        if (rand.nextInt(2) == 1) { // 0 - moving, 1 - not moving
            // world.getLogger().log("Turtle decided not to move\n");
            return 0;
        }

        int deltaY = ThreadLocalRandom.current().nextInt(-1, 2);; // Generates random number between -1 and 1
        int deltaX = ThreadLocalRandom.current().nextInt(-1, 2);; // Generates random number between -1 and 1

        if (deltaX == 0 && deltaY == 0) {
            return 0;
        }

        return move(new Point(deltaX, deltaY));
    }

    @Override
    public boolean rejectAttack(Organism attacker) {
        if (attacker.getStrength() < 5) {
            world.getLogger().log(LogManager.LogMessageType.ABILITY,"Turtle rejected attack");
            return true;
        }
        return false;
    }

    @Override
    public Organism makeChild(Point newPosition) {
        return new Turtle(world, newPosition);
    }
}

File: Sem2/PO/Project/C++/src/turtle.cpp
#include "turtle.h"
#include <cstdlib>

turtle::turtle(world* world_point, YX position)
:animal(2,1, position, world_point)
{
    avatar = 'T';
}

int turtle::Action() // 0 - normal move, 1 - collision detected
{
    if (rand() % 2 == 1) { // 0 - moving, 1 - not moving
        //world_point->GetLogger()->Log("turtle desided not to move\n");
        return 0;
    }

    int delta_y = rand() % 3 - 1; // Generates random number between -1 and 1
    int delta_x = rand() % 3 - 1; // Generates random number between -1 and 1

    if ((delta_x == 0) && (delta_y == 0)) {
        return 0;
    }

    return Move({delta_y, delta_x});
}

bool turtle::RejectAttak(organizm* attaker)
{
    if (attaker->GetStrength() < 5) {
        world_point->GetLogger()->Log("turtle rejected attak");
        return true;
    }

    return false;
}

organizm* turtle::MakeChild(YX new_position)
{
    return (new turtle(world_point, new_position));
}

turtle::~turtle()
{
}